src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt

```diff
@@ -563,28 +568,40 @@ class ObservationResultsStore(private val dslContext: DSLContext) {
       with(OBSERVED_PLOT_SPECIES_TOTALS) {
         speciesMultiset(
             DSL.select(
-                    CERTAINTY_ID,
+                    DSL.coalesce(CERTAINTY_ID, RecordedSpeciesCertainty.Known),
                     DSL.case_()
                         .`when`(OBSERVATION_PLOTS.IS_PERMANENT, MORTALITY_RATE)
                         .else_(null as Int?),
-                    SPECIES_ID,
+                    DSL.coalesce(
+                        SPECIES_ID,
+                        PLOT_T0_DENSITY.SPECIES_ID,
+                    ),
                     SPECIES_NAME,
-                    TOTAL_LIVE,
-                    TOTAL_DEAD,
-                    TOTAL_EXISTING,
-                    CUMULATIVE_DEAD,
-                    PERMANENT_LIVE,
-                    SURVIVAL_RATE,
-                    DSL.field(
-                        DSL.select(DSL.sum(PLOT_T0_DENSITY.PLOT_DENSITY))
-                            .from(PLOT_T0_DENSITY)
-                            .where(PLOT_T0_DENSITY.MONITORING_PLOT_ID.eq(MONITORING_PLOT_ID))
-                            .and(PLOT_T0_DENSITY.SPECIES_ID.eq(SPECIES_ID))
+                    DSL.coalesce(TOTAL_LIVE, 0),
+                    DSL.coalesce(TOTAL_DEAD, 0),
+                    DSL.coalesce(TOTAL_EXISTING, 0),
+                    DSL.coalesce(CUMULATIVE_DEAD, 0),
+                    DSL.coalesce(PERMANENT_LIVE, 0),
+                    DSL.coalesce(
+                        SURVIVAL_RATE,
+                        DSL.`when`(
+                            PLOT_T0_DENSITY.PLOT_DENSITY.isNotNull,
+                            DSL.inline(BigDecimal.ZERO),
+                        ),
                     ),
+                    PLOT_T0_DENSITY.PLOT_DENSITY,
                 )
                 .from(OBSERVED_PLOT_SPECIES_TOTALS)
-                .where(MONITORING_PLOT_ID.eq(MONITORING_PLOTS.ID))
-                .and(OBSERVATION_ID.eq(OBSERVATIONS.ID))
+                .fullOuterJoin(PLOT_T0_DENSITY)
+                .on(
+                    PLOT_T0_DENSITY.MONITORING_PLOT_ID.eq(MONITORING_PLOT_ID)
+                        .and(PLOT_T0_DENSITY.SPECIES_ID.eq(SPECIES_ID))
+                )
+                .where(
+                    MONITORING_PLOT_ID.eq(MONITORING_PLOTS.ID)
+                        .or(PLOT_T0_DENSITY.MONITORING_PLOT_ID.eq(MONITORING_PLOTS.ID))
+                )
+                .and(OBSERVATION_ID.eq(OBSERVATIONS.ID).or(OBSERVATION_ID.isNull))
```

Comment:
I think this approach is fine, but reading it, another possible approach occurred to me: using `UNION` and `DISTINCT ON` to split this up into two separate subqueries whose results get combined and deduplicated. I think it'd eliminate the need for the `COALESCE`s. But it might end up being more lines of code than this approach.

Reply:
(for reference) tested locally and it required a lot of jooq boilerplate. We decided to keep the original coalesce version.